### DISCUSSION TITLE
Periodic scrubbing

### DISCUSSION THREAD

**Discussion by Timmiej93** [openzfs/zfs#13136] [Q&A]
Hi,

I'm new to using zfs. I've installed it and set it up, everything works great. However - after watching Linus' (LTT) video about their data loss due to bit rot, I'd like to set up periodic scrubbing. For something which I'd consider so basic it should almost be turned on by default, I'm having a very hard time really finding any information about it. None of the tutorials online about ZFS even mention scrubbing, and I can't find much with Google either. So, two questions:

**Is periodic scrubbing really a good idea?**
**How do I enable periodic scrubbing?** (Preferably not through cron)

**Comment by rincebrain** [comment]
Periodic scrubbing is an exceedingly good idea, and common advice. Debian, for example, ships with it configured to trigger one every second Sunday of the month, I believe.

Stock OpenZFS ships a systemd timer you can use to trigger it periodically, though I don't recall offhand if it enables it by default. If you're not using systemd, then cron or some init-specific cron analogue would be my  suggestion.

**Reply by Timmiej93** [reply]
It does appear like I'm using systemd. I found [this page](https://wiki.archlinux.org/title/ZFS#Start_with_a_service_or_timer) now and followed that. I'm not sure if it worked though, since there's no feedback. 

`/etc/systemd/system/zfs-scrub@.timer`
```
[Unit]
Description=Monthly zpool scrub on %i

[Timer]
OnCalendar=monthly
AccuracySec=1h
Persistent=true

[Install]
WantedBy=multi-user.target
```

`/etc/systemd/system/zfs-scrub@.service`
```
[Unit]
Description=zpool scrub on %i

[Service]
Nice=19
IOSchedulingClass=idle
KillSignal=SIGINT
ExecStart=/usr/bin/zpool scrub %i

[Install]
WantedBy=multi-user.target
```

`sudo systemctl enable zfs-scrub@homepool` -> Created symlink /etc/systemd/system/multi-user.target.wants/zfs-scrub@homepool.service → /etc/systemd/system/zfs-scrub@.service.
`sudo systemctl start zfs-scrub@homepool` -> No feedback

Does this look correct? Is there any way for me to check if this has worked?

**Reply by rincebrain** [reply]
`systemctl status [timer name]`, I would presume.

**Reply by Timmiej93** [reply]
That's not promising:
```
● zfs-scrub@homepool.service - zpool scrub on homepool
     Loaded: loaded (/etc/systemd/system/zfs-scrub@.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Mon 2022-02-21 16:15:16 CET; 23min ago
    Process: 438549 ExecStart=/usr/bin/zpool scrub homepool (code=exited, status=203/EXEC)
   Main PID: 438549 (code=exited, status=203/EXEC)

feb 21 16:15:16 Quieter2 systemd[1]: Started zpool scrub on homepool.
feb 21 16:15:16 Quieter2 systemd[1]: zfs-scrub@homepool.service: Main process exited, code=exited, status=203/EXEC
feb 21 16:15:16 Quieter2 systemd[1]: zfs-scrub@homepool.service: Failed with result 'exit-code'.
```

**Reply by rincebrain** [reply]
That should have succeeded...is your zpool maybe not in /usr/bin? What happens if you try copy pasting the command from execstart directly?

(Though I did mean what you later did with status [the timer], not status [the service], triggering the service manually should still work, I believe, since I think that's all I expect the timer to do on trigger...)

**Reply by Timmiej93** [reply]
Interesting, zpool is in /usr/sbin/ for me. I must admit I didn't really look at the contents of the two examples they posted. Anyway, gmelikov posted the service and timer that are going to be included in the next major stable build of zfs, I copy pasted those and they seem to work. Not 100% sure, but at least it appears so.

**Reply by rincebrain** [reply]
I wouldn't expect the things he pasted to work without modification - the .in suffix means they're intended to be postprocessed during the build, and e.g. zfs-scrub...service.in contains a variable substitution for sbindir that you'd need to fill in, unless systemd evaluates that (which would surprise me).

**Reply by Timmiej93** [reply]
I noticed that too, so I did away with the .in extension. I also noticed the sbin variable, I just figured that would get taken care of. Just to be safe, I changed it to /usr/sbin.

I also had to take out the -w option on line 13. For some reason it threw an error as unrecognized, even though the option is in the documentation.

**Reply by rincebrain** [reply]
Which version are you on? It might indeed not be present.

**Reply by Timmiej93** [reply]
I just installed it very recently, so I'd assume it's the latest version: 
```
zfs-0.8.3-1ubuntu12.13
zfs-kmod-2.0.6-1ubuntu2
```

**Reply by rincebrain** [reply]
You appear to be running the kernel version from one Ubuntu release with the userland from another.

Thanks, Ubuntu HWE. You're such a pal.

That would help explain why - 0.8.X doesn't have that feature, and Ubuntu doesn't ship a package with 2.0.X userland for versions that ship 0.8.X kernel modules originally, AFAIK.

My general advice for people using Ubuntu is to use jonathonf's PPA or to build it themselves.

**Reply by Timmiej93** [reply]
I understood some of those words, lol. Thanks for all the help!

**Comment by gmelikov** [comment]
You may want to look at 
https://github.com/openzfs/zfs/blob/master/etc/systemd/system/zfs-scrub%40.service.in
https://github.com/openzfs/zfs/blob/master/etc/systemd/system/zfs-scrub-monthly%40.timer.in
https://github.com/openzfs/zfs/blob/master/etc/systemd/system/zfs-scrub-weekly%40.timer.in

It was added recently and will be shipped with next major stable version https://github.com/openzfs/zfs/commit/2300621dc7450adccf86ddbf397c72f8e5e2d76c

**Reply by Timmiej93** [reply]
I think that worked. Output from `sudo systemctl status zfs-scrub@homepool.timer`:
```
● zfs-scrub@homepool.timer - Monthly zpool scrub timer for homepool
     Loaded: loaded (/etc/systemd/system/zfs-scrub@.timer; enabled; vendor preset: enabled)
     Active: active (waiting) since Mon 2022-02-21 16:43:12 CET; 44s ago
    Trigger: Tue 2022-03-01 00:46:55 CET; 1 weeks 0 days left
   Triggers: ● zfs-scrub@homepool.service
       Docs: man:zpool-scrub(8)

feb 21 16:43:12 Quieter2 systemd[1]: Started Monthly zpool scrub timer for homepool.
```

No errors, appears to be waiting, that should work right?
I assume the service is only called when the timer triggers, so it's alright if that still shows failed?

**Comment by tomchiverton** [comment]
I have zfs-2.1.12-1.fc38.x86_64 and a pool

```
# zfs list
NAME           USED  AVAIL     REFER  MOUNTPOINT
tank          1.26M  3.51T      104K  /tank
tank/exports    96K  3.51T       96K  /tank/exports
tank/home       96K  3.51T       96K  /tank/home
```

This has provided 

```
# ls /usr/lib/systemd/system/zfs-scrub*
/usr/lib/systemd/system/zfs-scrub-monthly@.timer  /usr/lib/systemd/system/zfs-scrub-weekly@.timer
/usr/lib/systemd/system/zfs-scrub@.service
```
but there don't seem to be running, or listed in `systemctl list-timers`. 

Should something be happening to use these to generate ones named for each pool ?

**Comment by tomchiverton** [comment]
Edit: `systemctl enable zfs-scrub-weekly@tank.timer --now` works
